File: lib/discord.ts
import {
  Client,
  GatewayIntentBits,
  REST,
  Routes,
  SlashCommandBuilder,
  ChatInputCommandInteraction,
  AttachmentBuilder,
  Interaction,
} from 'discord.js';
import os from 'os';
import fs from 'fs';
import path from 'path';
import { randomUUID } from 'crypto';
import { getSetting } from './db';
import { runYtdlp } from './ytdlp';
import { runGalleryDl } from './gallerydl';
import { logger } from './logger';

// Discord's maximum attachment size (10 MB for standard servers)
const DISCORD_SIZE_LIMIT = 10 * 1024 * 1024;

let client: Client | null = null;

export async function startDiscordBot(): Promise<void> {
  const enabled = getSetting('discord_enabled');
  if (enabled !== 'true') return;

  const token = getSetting('discord_bot_token') ?? '';
  const clientId = getSetting('discord_client_id') ?? '';
  const commandName = getSetting('discord_command_name') || 'get';

  if (!token || !clientId) {
    logger.warn('Discord bot enabled but token or client ID not configured — skipping');
    return;
  }

  try {
    await registerCommand(token, clientId, commandName);
  } catch (err) {
    logger.error(`Discord: failed to register slash command: ${err instanceof Error ? err.message : err}`);
    return;
  }

  client = new Client({ intents: [GatewayIntentBits.Guilds] });

  client.once('clientReady', (c) => {
    logger.info(`Discord bot ready: ${c.user.tag}`);
  });

  client.on('interactionCreate', (interaction: Interaction) => {
    if (!interaction.isChatInputCommand()) return;
    if (interaction.commandName !== commandName) return;
    handleCommand(interaction).catch((err) => {
      logger.error(`Discord: unhandled error in command handler: ${err instanceof Error ? err.message : err}`);
    });
  });

  client.on('error', (err) => {
    logger.error(`Discord client error: ${err.message}`);
  });

  try {
    await client.login(token);
  } catch (err) {
    logger.error(`Discord: failed to log in: ${err instanceof Error ? err.message : err}`);
    client = null;
  }
}

export async function restartDiscordBot(): Promise<void> {
  if (client) {
    try {
      await client.destroy();
    } catch {
      // ignore errors during destroy
    }
    client = null;
  }
  await startDiscordBot();
}

async function registerCommand(token: string, clientId: string, commandName: string): Promise<void> {
  const command = new SlashCommandBuilder()
    .setName(commandName)
    .setDescription('Download a video or image and post it here')
    .addStringOption((opt) =>
      opt.setName('url').setDescription('The URL to download').setRequired(true)
    )
    .addStringOption((opt) =>
      opt
        .setName('type')
        .setDescription('Type of media to download (default: video)')
        .setRequired(false)
        .addChoices({ name: 'Video', value: 'video' }, { name: 'Image / Album', value: 'image' })
    );

  const rest = new REST().setToken(token);
  await rest.put(Routes.applicationCommands(clientId), { body: [command.toJSON()] });
  logger.info(`Discord: slash command "/${commandName}" registered`);
}

async function handleCommand(interaction: ChatInputCommandInteraction): Promise<void> {
  const url = interaction.options.getString('url', true);
  const type = (interaction.options.getString('type') ?? 'video') as 'video' | 'image';

  await interaction.deferReply();

  const tmpDir = path.join(os.tmpdir(), `memevault-discord-${randomUUID()}`);

  try {
    if (type === 'video') {
      const result = await runYtdlp(url, () => {}, undefined, tmpDir);
      const stat = fs.statSync(result.filePath);

      if (stat.size > DISCORD_SIZE_LIMIT) {
        const sizeMb = (stat.size / 1024 / 1024).toFixed(1);
        await interaction.editReply(
          `The file is too large to post (${sizeMb} MB). Discord's limit is 10 MB.`
        );
        return;
      }

      await interaction.editReply({ files: [new AttachmentBuilder(result.filePath)] });
    } else {
      const results = await runGalleryDl(url, () => {}, undefined, tmpDir);

      if (results.length === 0) {
        await interaction.editReply('No files were downloaded from that URL.');
        return;
      }

      const valid = results.filter((r) => fs.statSync(r.filePath).size <= DISCORD_SIZE_LIMIT);
      const oversizedCount = results.length - valid.length;

      if (valid.length === 0) {
        await interaction.editReply(
          `All ${results.length} file(s) exceed Discord's 10 MB limit and cannot be posted.`
        );
        return;
      }

      // Post in batches of 10 (Discord's per-message attachment limit)
      const BATCH_SIZE = 10;
      let firstReply = true;

      for (let i = 0; i < valid.length; i += BATCH_SIZE) {
        const batch = valid.slice(i, i + BATCH_SIZE);
        const attachments = batch.map((r) => new AttachmentBuilder(r.filePath));

        if (firstReply) {
          await interaction.editReply({ files: attachments });
          firstReply = false;
        } else {
          await interaction.followUp({ files: attachments });
        }
      }

      if (oversizedCount > 0) {
        await interaction.followUp(
          `${oversizedCount} file(s) were skipped because they exceed Discord's 10 MB limit.`
        );
      }
    }
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    logger.error(`Discord: download error for url=${url}: ${message}`);
    try {
      const logBuffer = Buffer.from(message, 'utf-8');
      const logAttachment = new AttachmentBuilder(logBuffer, { name: 'error.log' });
      await interaction.editReply({
        content: "Oops, seems like that didn't work!",
        files: [logAttachment],
      });
    } catch {
      // interaction may have already been replied to or expired
    }
  } finally {
    fs.rmSync(tmpDir, { recursive: true, force: true });
  }
}

File: src/discord.rs
//! Optional Discord slash-command bot (downloads to Discord, not vault).

use crate::config::Config;
use crate::db::{self, Db};
use crate::gallerydl;
use crate::state::{AppState, DiscordControl};
use crate::ytdlp;
use serenity::all::*;
use tokio::sync::watch;

pub async fn restart_discord_bot(state: AppState) {
    // Stop previous
    {
        let mut guard = state.discord.lock().await;
        if let Some(ctrl) = guard.take() {
            let _ = ctrl.shutdown.send(true);
        }
    }

    let enabled = state
        .db
        .with_conn(|c| Ok(db::get_setting(c, "discord_enabled")))
        .ok()
        .flatten()
        .as_deref()
        == Some("true");
    let token = state
        .db
        .with_conn(|c| Ok(db::get_setting(c, "discord_bot_token")))
        .ok()
        .flatten()
        .filter(|s| !s.is_empty());
    let client_id = state
        .db
        .with_conn(|c| Ok(db::get_setting(c, "discord_client_id")))
        .ok()
        .flatten()
        .filter(|s| !s.is_empty());

    if !enabled || token.is_none() || client_id.is_none() {
        tracing::info!("Discord bot not started (disabled or missing credentials)");
        return;
    }

    let token = token.unwrap();
    let (shutdown_tx, shutdown_rx) = watch::channel(false);
    {
        let mut guard = state.discord.lock().await;
        *guard = Some(DiscordControl {
            shutdown: shutdown_tx,
        });
    }

    let db = state.db.clone();
    let config = state.config.clone();
    let command_name = state
        .db
        .with_conn(|c| Ok(db::get_setting(c, "discord_command_name")))
        .ok()
        .flatten()
        .filter(|s| !s.is_empty())
        .unwrap_or_else(|| "get".into());

    tokio::spawn(async move {
        if let Err(e) = run_bot(token, command_name, db, config, shutdown_rx).await {
            tracing::error!("Discord bot error: {e:#}");
        }
    });
}

async fn run_bot(
    token: String,
    command_name: String,
    db: Db,
    config: Config,
    mut shutdown: watch::Receiver<bool>,
) -> anyhow::Result<()> {
    struct Handler {
        command_name: String,
        db: Db,
        config: Config,
    }

    #[async_trait::async_trait]
    impl EventHandler for Handler {
        async fn ready(&self, ctx: Context, ready: Ready) {
            tracing::info!("Discord bot logged in as {}", ready.user.name);
            let cmd = CreateCommand::new(&self.command_name)
                .description("Download media and post it here")
                .add_option(
                    CreateCommandOption::new(CommandOptionType::String, "url", "Media URL")
                        .required(true),
                )
                .add_option(
                    CreateCommandOption::new(CommandOptionType::String, "type", "video or image")
                        .add_string_choice("video", "video")
                        .add_string_choice("image", "image"),
                );
            if let Err(e) = Command::create_global_command(&ctx.http, cmd).await {
                tracing::error!("failed to register slash command: {e}");
            }
        }

        async fn interaction_create(&self, ctx: Context, interaction: Interaction) {
            let Interaction::Command(cmd) = interaction else {
                return;
            };
            if cmd.data.name != self.command_name {
                return;
            }
            let url = cmd
                .data
                .options
                .iter()
                .find(|o| o.name == "url")
                .and_then(|o| o.value.as_str())
                .unwrap_or("")
                .to_string();
            let media_type = cmd
                .data
                .options
                .iter()
                .find(|o| o.name == "type")
                .and_then(|o| o.value.as_str())
                .unwrap_or("video");

            let _ = cmd.defer(&ctx.http).await;
            let tmp = tempfile::tempdir().ok();
            let Some(tmp) = tmp else {
                let _ = cmd
                    .edit_response(
                        &ctx.http,
                        EditInteractionResponse::new().content("Failed to create temp dir"),
                    )
                    .await;
                return;
            };

            let (_, cancel_rx) = watch::channel(false);
            let result = if media_type == "image" {
                gallerydl::run_gallery_dl(
                    &self.db,
                    &self.config,
                    &url,
                    None,
                    cancel_rx,
                    Some(tmp.path().to_path_buf()),
                )
                .await
                .map(|files| {
                    files
                        .into_iter()
                        .map(|f| f.file_path)
                        .collect::<Vec<_>>()
                })
            } else {
                ytdlp::run_ytdlp(
                    &self.db,
                    &self.config,
                    &url,
                    None,
                    cancel_rx,
                    Some(tmp.path().to_path_buf()),
                )
                .await
                .map(|r| vec![r.file_path])
            };

            match result {
                Ok(paths) if !paths.is_empty() => {
                    let _ = cmd
                        .edit_response(
                            &ctx.http,
                            EditInteractionResponse::new()
                                .content(format!("Downloaded {} file(s) from <{url}>", paths.len())),
                        )
                        .await;
                    for p in paths.iter().take(10) {
                        if let Ok(att) = CreateAttachment::path(p).await {
                            let _ = cmd
                                .create_followup(
                                    &ctx.http,
                                    CreateInteractionResponseFollowup::new().add_file(att),
                                )
                                .await;
                        }
                    }
                }
                Ok(_) => {
                    let _ = cmd
                        .edit_response(
                            &ctx.http,
                            EditInteractionResponse::new().content("No files downloaded"),
                        )
                        .await;
                }
                Err(e) => {
                    let _ = cmd
                        .edit_response(
                            &ctx.http,
                            EditInteractionResponse::new()
                                .content(format!("Download failed: {e}")),
                        )
                        .await;
                }
            }
        }
    }

    let handler = Handler {
        command_name,
        db,
        config,
    };
    let mut client = Client::builder(&token, GatewayIntents::empty())
        .event_handler(handler)
        .await?;

    tokio::select! {
        r = client.start() => { r?; }
        _ = shutdown.changed() => {
            tracing::info!("Discord bot shutting down");
        }
    }
    Ok(())
}

pub async fn start_if_configured(state: AppState) {
    restart_discord_bot(state).await;
}
